asvdb/asvdb.py

```diff
@@ -925,49 +938,191 @@ def __createLockfile(self, dirPath):
+                        if err not in e.response["Error"]["Message"]:
+                            raise
+
+                # Download specific result file for updating results if BenchmarkInfo is sent
+                try:
+                    if bInfo.machineName != "":
+                        commitHash, pyVer, cuVer, osType = bInfo.commitHash, bInfo.pythonVer, bInfo.cudaVer, bInfo.osType
+                        filename = f"{commitHash}-python{pyVer}-cuda{cuVer}-{osType}.json"
+                        os.makedirs(path.join(self.localS3Copy.name, self.defaultResultsDirName, bInfo.machineName), exist_ok=True)
+                        resultFileExt = path.join(self.defaultResultsDirName, bInfo.machineName, filename)
+                        downloadS3(bucket, resultFileExt)
+
+                except exceptions.ClientError as e:
+                    err = "Not Found"
+                    if err not in e.response["Error"]["Message"]:
+                        raise
+
+            else:
+                try:
+                    downloadS3(bucket, self.confFileExt)
+                except:
+                    err = "Not Found"
+                    if err not in e.response["Error"]["Message"]:
+                        raise
+
+                try:
+                    resultsBucketPath = path.join(self.bucketKey, self.defaultResultsDirName)
+                    resultsLocalPath = path.join(self.localS3Copy.name, self.defaultResultsDirName)
+                    
+                    # Loop over ASV results folder and download everything.
+                    # objectExt represents the file extension starting from the base resultsBucketPath
+                    # For example: resultsBucketPath = "asvdb/results"
+                    #            : objectKey = "asvdb/results/machine_name/results.json
+                    #            : objectExt = "machine_name/results.json"
+                    for bucketObj in bucket.objects.filter(Prefix=resultsBucketPath):
+                        objectExt = bucketObj.key.replace(resultsBucketPath + "/", "")
+                        if len(objectExt.split("/")) > 1:
+                            os.makedirs(path.join(resultsLocalPath, objectExt.split("/")[0]), exist_ok=True)
+                        bucket.download_file(bucketObj.key, path.join(resultsLocalPath, objectExt))
+                
+                except exceptions.ClientError as e:
+                    err = "Not Found"
+                    if err not in e.response["Error"]["Message"]:
+                        raise e
+            
+            # Set all the internal locations to point to the downloaded files:
+            self.confFilePath = path.join(self.localS3Copy.name, self.confFileName)
+            self.resultsDirPath = path.join(self.localS3Copy.name, self.resultsDirName)
+            self.benchmarksFilePath = path.join(self.resultsDirPath, self.benchmarksFileName)
 
 
     def __uploadIfS3(self):
+        def recursiveUpload(base, ext=""):
```

Comment:
This looks fine to me, but you might be interested in https://docs.python.org/3/library/os.html#os.walk which may allow you to remove a bunch of code.

Reply:
At first the implementation was using os.walk, but then I couldn't properly define the correct paths in S3 to upload the files. That led me to write this recursive function which used os.walk at first, but due to some weird directory structure interactions and how the function had to work, os.walk wasn't returning reliably.

If you'd like I could go back and show you why the approach wasn't working, but the os.walk return object wasn't giving me a consistent number of variables.